### PR TITLE
chore(authenticator): add additional types for form fields

### PR DESCRIPTION
**Pull request by thaddmt**
#### Description of changes
- Added additional types to form fields for `minLength`, `maxLength`, `pattern`, `defaultValue` and `isReadOnly`
- Fixes https://github.com/aws-amplify/amplify-ui/issues/1669
- Added note in docs to check out html form validation for additional validation standards

#### Checklist

- [ ] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [ ] PR description included
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] `yarn test` passes and tests are updated/added
- [ ] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by hbuchel — APPROVED** [review]
LGTM

**Review by reesscot — DISMISSED** [review]
LGTM

File: packages/ui/src/types/authenticator/form.ts [hunk 1]
```diff
@@ -64,6 +64,16 @@ export interface FormFieldOptions {
   type?: string;
   /** Desired autocomplete HTML attribute */
   autocomplete?: string;
+  /** Desired HTML pattern type */
+  pattern?: RegExp;
```
Comment (reesscot) [inline packages/ui/src/types/authenticator/form.ts:68]: Looks like you may want to pick this from Input type: https://github.com/aws-amplify/amplify-ui/actions/runs/5955430849/job/16154297954?pr=4376

File: packages/ui/src/types/authenticator/form.ts [hunk 2]
```diff
@@ -64,6 +64,16 @@ export interface FormFieldOptions {
   type?: string;
   /** Desired autocomplete HTML attribute */
   autocomplete?: string;
+  /** Desired HTML pattern type */
+  pattern?: string | undefined;
```
Comment (calebpollman) [inline packages/ui/src/types/authenticator/form.ts:68]: Not sure if this going to work as intended without typing as `RegExp` :
CleanShot 2023-08-24 at 09 06 07@2x

Comment (thaddmt) [reply packages/ui/src/types/authenticator/form.ts:68]: Hm I tried testing with a value like this and it seemed to work, open to other ideas here though!

```
given_name: {
      isRequired: true,
      label: 'custom name:',
      pattern: "[Bb]anana|[Cc]herry",
    },
```

Comment (thaddmt) [reply packages/ui/src/types/authenticator/form.ts:68]: But yes we will have an issue where you cannot pass a plain regexp here

Comment (calebpollman) [reply packages/ui/src/types/authenticator/form.ts:68]: Looked at the base input typing and are with just `string`

**Review by calebpollman — APPROVED** [review]

**Review by reesscot — APPROVED** [review]

File: examples/next/pages/ui/components/authenticator/custom-slots/index.page.tsx [hunk 3]
```diff
@@ -15,6 +15,7 @@ import {
 } from '@aws-amplify/ui-react';
 
 import awsExports from './aws-exports';
+import { AuthFormFields } from '@aws-amplify/ui';
```
Comment (calebpollman) [inline examples/next/pages/ui/components/authenticator/custom-slots/index.page.tsx:18]: Sorry missed this on the earlier revisions but can we have this example without the type? Would prefer we don't guide users to import from our internal packages

Comment (reesscot) [reply examples/next/pages/ui/components/authenticator/custom-slots/index.page.tsx:18]: +1, also means JS users can't just copy/past our examples.

File: packages/react/src/components/Authenticator/Authenticator.tsx [hunk 4]
```diff
@@ -36,9 +38,27 @@ export type AuthenticatorProps = Partial<
       children:
         | React.ReactNode
         | ((props: { signOut?: SignOut; user?: AmplifyUser }) => JSX.Element);
+      formFields: {
```
Comment (calebpollman) [inline packages/react/src/components/Authenticator/Authenticator.tsx:41]: Can we `Omit` the existing `formFields` prop on `AuthenticatorMachineOptions`?

File: packages/react/src/components/Authenticator/Authenticator.tsx [hunk 5]
```diff
@@ -4,6 +4,8 @@ import {
   AmplifyUser,
   configureComponent,
   isFunction,
+  FormFieldOptions,
```
Comment (calebpollman) [inline packages/react/src/components/Authenticator/Authenticator.tsx:7]: Think we should be using `LegacyFormFieldOptions` here

**Review by calebpollman — APPROVED** [review]
🚢

**Review by reesscot — APPROVED** [review]
LGTM

**Pull request merged**